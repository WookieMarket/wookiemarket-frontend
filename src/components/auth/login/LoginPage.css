.container-form {
  width: 50%;
  height: 50vh; /* Establece la altura del contenedor en el 50% de la altura de la ventana */
  position: absolute; /* Cambia a posici√≥n absoluta */
  top: 55%; /* Alinea el borde superior del contenedor al 50% de la ventana */
  left: 50%; /* Alinea el borde izquierdo del contenedor al 50% de la ventana */
  transform: translate(-50%, -50%); /* Centra vertical y horizontalmente */
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
  padding: 30px 10px 20px 10px;
  gap: 10px;
  border-radius: 10px;
  text-align: center;
  border: 2px solid var(--border-color);
  background-color: var(--background-color-form);
}

.container-form .form-group {
  width: 100%;
  padding: 0 15px;
  display: flex;
  flex-direction: column;
  gap: 10px;
}

.container-form .password-label {
  font-family: var(--font-family);
  font-size: 13px;
  margin-bottom: 10px;
}

.container-form .password-input {
  height: 20px;
  width: 100%;
  font-family: var(--font-family);
  margin: auto;
  font-size: 12px;
  margin-bottom: 10px;
  border-radius: 6px;
  border: 3px solid gainsboro;
}
.container-form .email-input {
  height: 20px;
  width: 100%;
  margin: auto;
  font-size: 13px;
  margin-bottom: 10px;
  border-radius: 6px;
  border: 3px solid gainsboro;
}

.container-form .link-reset {
  display: flex;
  justify-content: center;
  font-family: var(--font-family);
  font-size: 14px;
  color: rgba(0, 0, 255, 0.897);
  text-decoration: none;
  margin-bottom: 10px;
}

.container-form .password-input-rememberme {
  display: flex;
  font-family: var(--font-family);
  align-items: center;
  align-content: space-between;
  font-size: 14px;
}

@media screen and (min-width: 768px) {
  .container-form {
    width: 40%;
  }
  .container-form .password-input {
    height: 25px;
    font-size: 12px;
  }
  .container-form .password-input-rememberme {
    font-size: 14px;
  }
  /* .container-form .login-button {
    border-radius: 5px;
    padding: 7px 7px;
    font-size: 15px;
  } */
  .container-form .link-reset {
    font-size: 13px;
  }
  .container-form .password-label {
    font-size: 14px;
    margin-bottom: 10px;
  }
}

@media screen and (min-width: 1024px) {
  .container-form {
    width: 30%;
  }
  .container-form .password-input {
    height: 30px;
    font-size: 14px;
  }
  .container-form .link {
    font-size: 14px;
  }

  .container-form .password-input-rememberme {
    font-size: 14px;
  }
  /* .container-form .login-button {
    border-radius: 5px;
    padding: 8px 8px;
    font-size: 15px;
  } */
  .container-form .password-label {
    font-size: 16px;
    margin-bottom: 10px;
  }
}
