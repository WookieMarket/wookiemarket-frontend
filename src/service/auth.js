import client, {
  removeAuthorizationHeader,
  setAuthorizationHeader,
} from '../api/client';
import storage from '../utils/storage';

export const signup = userData => {
  return client.post('/api/auth/signup', userData).then(({ jwt }) => {
    setAuthorizationHeader(jwt);
    storage.set('auth', jwt);
  });
};

export const login = credentials => {
  return client.post('/api/auth/login', credentials).then(({ jwt }) => {
    setAuthorizationHeader(jwt);
    if (credentials.rememberMe) {
      storage.set('auth', jwt);
    }
    console.log('jwt,', jwt);
    return jwt;
  });
};

export const logout = () => {
  return Promise.resolve().then(() => {
    removeAuthorizationHeader();
    storage.remove('auth');
  });
};

export const emailPassword = email => {
  return client.post('api/users/email-password', { to: email });
};

export const resetPassword = (email, newPassword, token) => {
  return client.post('api/users/recover-password', {
    email,
    newPassword,
    token,
  });
};

export const deleteUser = email => {
  return client.post('api/users/deleted-user', { email: email });
};
