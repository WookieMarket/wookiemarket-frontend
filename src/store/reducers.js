export {default as adverts} from './slices/adverts';
export {default as ui} from './slices/ui';