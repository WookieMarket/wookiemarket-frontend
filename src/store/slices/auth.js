import { createAsyncThunk, createSlice } from '@reduxjs/toolkit';

// use rejectWithValue to take care of the error

export const authSignup = createAsyncThunk(
  'auth/signup',
  async (userData, { extra: { service }, rejectWithValue }) => {
    try {
      await service.auth.signup(userData);
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const authLogin = createAsyncThunk(
  'auth/login',
  async (credentials, { extra: { service }, rejectWithValue }) => {
    try {
      const response = await service.auth.login(credentials);
      return response;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const authLogout = createAsyncThunk(
  'auth/logout',
  (_, { extra: { service } }) => service.auth.logout(),
);

export const emailResetPassword = createAsyncThunk(
  'auth/emailResetPassword',
  async (email, { extra: { service }, rejectWithValue }) => {
    try {
      const resetEmail = await service.auth.emailPassword(email);
      return resetEmail;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const resetPassword = createAsyncThunk(
  'auth/resetPassword',
  async (
    { email, newPassword, token },
    { extra: { service }, rejectWithValue },
  ) => {
    try {
      const response = await service.auth.resetPassword(
        email,
        newPassword,
        token,
      );
      return response;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

export const deleteAccount = createAsyncThunk(
  'auth/deleteAccount',
  async (email, { extra: { service }, rejectWithValue }) => {
    try {
      const response = await service.auth.deleteUser(email);
      return response;
    } catch (error) {
      return rejectWithValue(error);
    }
  },
);

const auth = createSlice({
  name: 'auth',
  initialState: {
    auth: false,
    jwt: '',
  },
  extraReducers: builder => {
    builder
      .addCase(authSignup.fulfilled, (state, action) => {
        state.auth = true;
        state.jwt = '';
      })
      .addCase(authLogin.fulfilled, (state, action) => {
        state.auth = true;
        state.jwt = action.payload;
      })
      .addCase(authLogout.fulfilled, (state, action) => {
        state.auth = false;
        state.jwt = '';
      })
      .addCase(deleteAccount.fulfilled, (state, action) => {
        state.auth = false;
        state.jwt = '';
      });
  },
});

export default auth.reducer;
