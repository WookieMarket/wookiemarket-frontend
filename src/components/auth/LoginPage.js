import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getUi } from "../../store/selectors";
import { resetError } from "../../store/slices/ui";
import { authLogin } from "../../store/slices/auth";
import ErrorModal from "../shared/modal/ErrorModal";
import Spinner from "../shared/spinner/Spinner";
import Layout from "../layout/Layout";

//DONE Log in with username and password and a checkbox to give the option to persist the token, also handle errors and user feedback. When doing Login I want to send the user to the page they wanted to go to.

function LoginPage() {
  const dispatch = useDispatch();
  const { isLoading, error } = useSelector(getUi);

  const [credentials, setCredentials] = useState({
    username: "",
    password: "",
    rememberMe: false,
  });

  const handleErrorClick = () => {
    dispatch(resetError());
  };

  const handleChange = event => {
    const { name, value, type, checked } = event.target;

    setCredentials(prevCredentials => ({
      ...prevCredentials,
      [name]: type === "checkbox" ? checked : value,
    }));
  };

  const handleSubmit = async event => {
    event.preventDefault();

    dispatch(authLogin(credentials));
  };

  const buttonDisabled =
    isLoading || !credentials.username || !credentials.password;

  return (
    <Layout title="Login Page">
      <div>
        {isLoading ? (
          <Spinner message="cargando..." />
        ) : (
          <form onSubmit={handleSubmit} className="container-form">
            <label className="form-label" htmlFor="username">
              Username
            </label>
            <input
              id="username"
              className="form-input"
              type="text"
              name="username"
              onChange={handleChange}
              value={credentials.username}
              required
            />
            <label className="form-label" htmlFor="password">
              Password
            </label>
            <input
              id="password"
              className="form-input"
              type="password"
              name="password"
              onChange={handleChange}
              value={credentials.password}
              required
            />
            <label htmlFor="rememberMe">
              rememberMe
              <input
                id="rememberMe"
                type="checkbox"
                name="rememberMe"
                checked={credentials.rememberMe}
                onChange={handleChange}
              />
            </label>
            <button
              data-testid="button"
              type="submit"
              //variant="primary"
              width="button-form"
              disabled={buttonDisabled}>
              Log in
            </button>
          </form>
        )}

        {error && (
          <ErrorModal
            title="Error"
            message={error.message}
            onCancel={handleErrorClick}
          />
        )}
      </div>
    </Layout>
  );
}

export default LoginPage;
