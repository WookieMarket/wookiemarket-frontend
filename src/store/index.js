import { configureStore as rtkConfigureStore } from '@reduxjs/toolkit';
import * as service from '../service';
import * as reducers from './reducers';
import { failureRedirects, successRedirects } from './middleware';
import { authSignup, authLogin } from './slices/auth';

export default function configureStore(preloadedState, { router }) {
  const extraMiddleware = [
    failureRedirects(router, { 401: '/login', 404: '/404' }),
    successRedirects(router, {
      [authSignup.fulfilled.type]: () => {
        return router.state.location.state?.from?.pathname || '/';
      },
      [authLogin.fulfilled.type]: () => {
        return router.state.location.state?.from?.pathname || '/';
      },
    }),
  ];

  const store = rtkConfigureStore({
    reducer: reducers,
    middleware: getDefaultMiddleware =>
      getDefaultMiddleware({
        thunk: { extraArgument: { service, router } },
        serializableCheck: false,
      }).concat(extraMiddleware),
    preloadedState,
  });
  return store;
}
