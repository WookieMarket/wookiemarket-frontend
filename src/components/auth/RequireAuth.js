import { useSelector } from 'react-redux';
import { getIsLogged } from '../../store/selectors';
import { Navigate, useLocation } from 'react-router-dom';

//NOTE The children property refers to the element that is inside the tag

//DONE Function that I use to cover a component if it is required to have authorization, I pass a location object through the state property to know which route I come from

const RequireAuth = ({ children }) => {
  const isLogged = useSelector(getIsLogged);
  const location = useLocation();

  if (!isLogged) {
    return <Navigate to="/login" state={{ from: location }} />;
  }
  return children;
};

export default RequireAuth;
