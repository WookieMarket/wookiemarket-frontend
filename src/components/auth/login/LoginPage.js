import { useState } from "react";
import { useDispatch, useSelector } from "react-redux";
import { getUi } from "../../../store/selectors";
import { resetError, toggleModal } from "../../../store/slices/ui";
import { authLogin, emailResetPassword } from "../../../store/slices/auth";
import ErrorModal from "../../shared/modal/ErrorModal";
import Spinner from "../../shared/spinner/Spinner";
import Layout from "../../layout/Layout";
import { useTranslation } from "react-i18next";
import Modal from "../../shared/modal/Modal";
import { Link } from "react-router-dom";
import Form from "../../shared/form/Form";
import "./LoginPage.css";
import Button from "../../shared/Button";

//DONE Log in with username and password and a checkbox to give the option to persist the token, also handle errors and user feedback. When doing Login I want to send the user to the page they wanted to go to.

function LoginPage() {
  const { t } = useTranslation();
  const dispatch = useDispatch();
  const { isLoading, error, showModal } = useSelector(getUi);

  const [credentials, setCredentials] = useState({
    username: "",
    password: "",
    rememberMe: false,
  });

  const [email, setEmail] = useState("");

  const handleErrorClick = () => {
    dispatch(resetError());
  };

  const handleChange = event => {
    const { name, value, type, checked } = event.target;

    setCredentials(prevCredentials => ({
      ...prevCredentials,
      [name]: type === "checkbox" ? checked : value,
    }));
  };

  const handleSubmit = async event => {
    event.preventDefault();

    dispatch(authLogin(credentials));
  };

  const handleResetPasswordLinkClick = () => {
    dispatch(toggleModal());
  };

  const handleShowModalconfirm = async event => {
    event.preventDefault();
    dispatch(emailResetPassword(email));
    dispatch(toggleModal());
  };

  const handleShowModalCancel = () => {
    dispatch(toggleModal());
  };

  const handleChangeEmail = e => {
    setEmail(e.target.value);
  };

  const buttonDisabled =
    isLoading || !credentials.username || !credentials.password;

  return (
    <Layout title={t("Login Page")}>
      <div>
        {isLoading ? (
          <Spinner message={t("charging...")} />
        ) : (
          <form onSubmit={handleSubmit} className="container-form">
            <Form
              classNameForm="form-group"
              classNameLabel="password-label"
              htmlFor="username"
              text={t("Username")}
              classNameInput="password-input"
              inputId="username"
              inputType="text"
              inputName="username"
              value={credentials.username}
              handleChange={handleChange}
              placeholder={t("Username")}
              required
            />

            <Form
              classNameForm="form-group"
              classNameLabel="password-label"
              htmlFor="password"
              text={t("Password")}
              classNameInput="password-input"
              inputId="password"
              inputType="password"
              inputName="password"
              value={credentials.password}
              handleChange={handleChange}
              placeholder={t("Password")}
              required
            />

            <Form
              classNameForm="password-input-rememberme"
              htmlFor="rememberMe"
              text={t("RememberMe")}
              inputId="rememberMe"
              inputType="checkbox"
              inputName="rememberMe"
              checked={credentials.rememberMe}
              handleChange={handleChange}
              required
            />

            <Button
              //className="login-button"
              data-testid="button"
              type="submit"
              variant="accept"
              width="button-form"
              disabled={buttonDisabled}>
              {t("Log in")}
            </Button>
            <Link onClick={handleResetPasswordLinkClick}>
              <h4 className="link-reset">{t("Forgot password?")}</h4>
            </Link>
            {showModal && (
              <Modal
                title={t("Recover password")}
                message={
                  <Form
                    classNameForm="form-group"
                    classNameLabel="password-label"
                    htmlFor="email-service"
                    text={t("Enter your email")}
                    classNameInput="email-input"
                    inputId="email-service"
                    inputType="email"
                    inputName="reset-email"
                    value={email}
                    handleChange={handleChangeEmail}
                    placeholder={t("Email")}
                    required
                  />
                }
                onConfirm={handleShowModalconfirm}
                onCancel={handleShowModalCancel}
              />
            )}
          </form>
        )}

        {error && (
          <ErrorModal
            title="Error"
            message={error.message}
            onCancel={handleErrorClick}
          />
        )}
      </div>
    </Layout>
  );
}

export default LoginPage;
