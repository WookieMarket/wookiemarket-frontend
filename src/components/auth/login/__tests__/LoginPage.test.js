import { render, screen } from "@testing-library/react";
import LoginPage from "../LoginPage";
import userEvent from "@testing-library/user-event";
import { authLogin } from "../../../../store/slices/auth";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import * as defaultState from "../../../../store/reducers";

jest.mock("../../../../store/slices/auth", () => ({
  __esModule: true,
  authLogin: jest.fn(), // Mock only the authLogin action
}));

describe("LoginPage", () => {
  const renderComponent = (error = null) => {
    const store = {
      getState: () => {
        const state = { ...defaultState }; // Create a copy of the state
        state.ui.error = error; // Modify the copy instead of the original state

        return state;
      },
      subscribe: () => {},
      dispatch: () => {},
    };
    return render(
      <Provider store={store}>
        <MemoryRouter>
          <LoginPage />
        </MemoryRouter>
      </Provider>,
    );
  };

  test("snapshot", () => {
    const { container } = renderComponent();
    expect(container).toMatchSnapshot();
  });

  test("shoul dispatch authLogin action", () => {
    const credentials = {
      username: "rober",
      password: "123",
      rememberMe: true,
    };

    //NOTE renderizo el componente
    renderComponent();
    const usernameInput = screen.getByLabelText(/Username/);
    const passwordInput = screen.getByLabelText(/Password/);
    const checkboxInput = screen.getByLabelText(/RememberMe/);
    const submitButton = screen.getByRole("button", { name: /Log in/ });
    expect(submitButton).toBeDisabled();

    //NOTE para lanzar eventos
    userEvent.type(usernameInput, credentials.username);
    userEvent.type(passwordInput, credentials.password);
    userEvent.click(checkboxInput, credentials.rememberMe);

    expect(submitButton).toBeEnabled();
    userEvent.click(submitButton);

    expect(authLogin).toHaveBeenCalledWith(credentials);
  });
  test("should display an error", () => {
    //NOTE // Spy on resetError function
    const resetErrorSpy = jest.spyOn(
      require("../../../../store/slices/ui"),
      "resetError",
    );

    const error = { message: "Unauthorized" };

    renderComponent(error);
    const errorElement = screen.getByText(error.message);

    expect(errorElement).toBeInTheDocument();
    const modalButton = screen.getByTestId("modalButton");
    userEvent.click(modalButton);

    expect(resetErrorSpy).toHaveBeenCalled();
  });
});
