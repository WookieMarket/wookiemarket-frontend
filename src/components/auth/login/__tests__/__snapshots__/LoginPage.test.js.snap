// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginPage snapshot 1`] = `
<div>
  <div>
    <header
      class="header"
    >
      <nav
        class="navbar"
      >
        <div
          class="icons"
        >
          <a
            href="/"
          >
            <h1
              class="title"
            >
               
              header
               
            </h1>
          </a>
          <img
            alt="ES"
            class="icon-language"
            src="/images/languageIcons/es.png"
            title="icon-hover"
          />
          <img
            alt="EN"
            class="icon-language"
            src="/images/languageIcons/en.png"
            title="icon-hover"
          />
        </div>
        <ul>
          <li>
            <a
              class="navbar-list-item"
              href="/create-ad"
            >
              Upload ad
            </a>
            <a
              class="navbar-list-item"
              href="/ads"
            >
              Ads
            </a>
            <a
              class="navbar-list-item"
              href="/login"
            >
              Login
            </a>
          </li>
        </ul>
      </nav>
      <hr />
    </header>
    <main
      class="content"
    >
      <h2
        class="layout-title"
      >
        Login Page
      </h2>
      <div>
        <form
          class="container-form"
        >
          <div
            class="form-group"
          >
            <label
              class="password-label"
              for="username"
            >
              Username
            </label>
            <input
              class="password-input"
              id="username"
              name="username"
              placeholder="Username"
              type="text"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label
              class="password-label"
              for="password"
            >
              Password
            </label>
            <input
              class="password-input"
              id="password"
              name="password"
              placeholder="Password"
              type="password"
              value=""
            />
          </div>
          <div
            class="password-input-rememberme"
          >
            <label
              for="rememberMe"
            >
              RememberMe
            </label>
            <input
              id="rememberMe"
              name="rememberMe"
              type="checkbox"
              value=""
            />
          </div>
          <button
            class="sc-bcPKhP ibftXS"
            data-testid="button"
            disabled=""
            type="submit"
            variant="primary"
            width="button-form"
          >
            Log in
          </button>
          <a
            href="/"
          >
            <h4
              class="link-reset"
            >
              Forgot password?
            </h4>
          </a>
        </form>
      </div>
    </main>
    <footer>
      <hr />
      <div
        class="footer"
      >
        <p>
          footer
        </p>
      </div>
    </footer>
  </div>
</div>
`;
